reinforcement-learning/dqn.py

```diff
@@ -0,0 +1,317 @@
+
+
+# You can find your API token at https://sigopt.com/user/tokens
+SIGOPT_API_KEY = 'YOUR_API_TOKEN_HERE'
+# The game to train the DQN on. The default is CartPole-v0. You can also try LunarLander-v0 and Acrobot-v1
+ENVIRONMENT_NAME = 'CartPole-v0'
+# These values depend on which game you play. Choose reasonable values so that the agent has sufficient time to learn
+MAX_EPISODES = 350
+MAX_TIMESTEPS = 10000
+# Determines whether the game should be rendered - set this to True to watch the agent play
+SHOULD_RENDER = False
+
+
+def main():
+  conn = sigopt.Connection(client_token=SIGOPT_API_KEY)
+
+  experiment = conn.experiments().create(
+    name= ENVIRONMENT_NAME + ' (DQN)',
+    observation_budget=128,
+    parameters=[
+      dict(name='hm', type='int', bounds=dict(min=5, max=100)),
+      dict(name='mb_size', type='int', bounds=dict(min=10, max=500)),
+      dict(name='e_decay', type='int', bounds=dict(min=MAX_EPISODES/10, max=MAX_EPISODES)),
+      dict(name='log_lr', type='double', bounds=dict(min=math.log(0.00001), max=math.log(1.0))),
+      dict(name='df', type='double', bounds=dict(min=0.5, max=0.9999)),
+      dict(name='weight_sd', type='double', bounds=dict(min=0.01, max=0.5)),
+      dict(name='bias_sd', type='double', bounds=dict(min=0.0, max=0.5))
+    ]
+  )
+
+  # Run the SigOpt experiment loop
+  for _ in range(experiment.observation_budget):
+    suggestion = conn.experiments(experiment.id).suggestions().create()
+
+    objective_metric = run_environment(
+      hidden_multiplier=suggestion.assignments['hm'],
+      minibatch_size=suggestion.assignments['mb_size'],
+      epsilon_decay_steps=suggestion.assignments['e_decay'],
+      learning_rate=math.exp(suggestion.assignments['log_lr']),
+      discount_factor=suggestion.assignments['df'],
+      initial_weight_stddev=suggestion.assignments['weight_sd'],
+      initial_bias_stddev=suggestion.assignments['bias_sd']
+    )
+
+    conn.experiments(experiment.id).observations().create(
+      suggestion=suggestion.id,
+      value=objective_metric,
+    )
+
+
+def run_environment(
+  hidden_multiplier,
+  minibatch_size,
+  epsilon_decay_steps,
+  learning_rate,
+  discount_factor,
+  initial_weight_stddev,
+  initial_bias_stddev
+):
+
+  env = gym.make(ENVIRONMENT_NAME)
+  # The number of states about the environment the agent can observe
+  observation_space_dim = env.observation_space.shape[0]
+  # The number of actions the agent can make
+  action_space_dim = env.action_space.n
+
+  # The dimensions of the layers of the Q-network
+  layer_dims = [
+    observation_space_dim,
+    hidden_multiplier * observation_space_dim,
+    hidden_multiplier * observation_space_dim,
+    action_space_dim
+  ]
+
+  neural_net = QNetwork(
+    layer_dims,
+    dict(
+      learning_rate=learning_rate,
+      initial_weight_stddev=initial_weight_stddev,
+      initial_bias_stddev=initial_bias_stddev
+    )
+  )
+
+  with tf.Session() as session:
+    agent = Agent(
```

Comment:
Since this is already long enough to break into multiple lines it's probably worth doing:

```python
Agent(
  session=session,
  neural_net=neural_net,
  ...
)
```

which helps make the code less brittle in the event of future changes